### PR TITLE
Fixed a deadlock when printing surrogate pairs

### LINKED ISSUE #1360: Windows terminal hangs after running ripgrep in a loop
# Environment
Windows version 1903 OS build 18899.1000
Windows Terminal (Preview) version 0.2.1703.0

Any other software?
ripgrep 11.0.0

# Steps to reproduce
Install [ripgrep](https://github.com/BurntSushi/ripgrep)
Open wt (Windows Terminal), running PowerShell.
Run the following from the root of a large drive
`while ($true) { rg -i "foo" }`

# Expected behavior
The terminal continually runs ripgrep against all files on the drive

# Actual behavior
The terminal eventually hangs and is completely unresponsive.

The call stack for the main thread when the window hangs is attached.

[HungTerminal1.callstack.txt](https://github.com/microsoft/terminal/files/3312675/HungTerminal1.callstack.txt)

A minidump of the hung process is available here:
[HungWindowsTerminal](https://1drv.ms/u/s!Alav93a0o3vchOg0nCVjwUPjFn0DYw?e=Ue4AP5)

### LINKED ISSUE #4145: Terminal crashes when WSL tab tries to display some characters.
Terminal (0.7.3451.0) crashes every time I try to cat or curl a html file containing a certain tweet. The entire terminal frame hangs, white out, goes to (Not Responding).
This happens only with the WSL Ubuntu tab and not when I try to cat the same file with powershell in Terminal. It does not crash with WSL under the older 'Ubuntu' launcher from the Windows Store.

# Environment
Windows 10 Pro 1909
Win32NT	10.0.18363.0	Microsoft Windows NT 10.0.18363.0

Windows Terminal (Preview) Version: 0.7.3451.0 from the Windows Store
Ubuntu 18.04.3 LTS installed from the Windows Store

# Steps to reproduce
WSL Ubuntu from the Windows Store installed, Terminal from Windows Store installed
Launch Windows Terminal
Window chrome down arrow: Ubuntu

curl 'https://twitter.com/EU_Commission/status/1213399676201734146' -H 'User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:71.0) Gecko/20100101 Firefox/71.0'

curl starts to display the html from that url, but doesn't complete as the whole window frame hangs at some point during the output 

# Expected behavior
I expected the whole output from curl to be displayed, and return to the bash prompt

# Actual behavior
curl starts displaying html output, but then the whole app frame including all tabs becomes unresponsive.

Although Windows displayed the "sending data to Microsoft" from the crash, I don't have a link to share as this computer has no Microsoft Account associated with it, and the feedback tool appears to require one to view crash data.

### PR DESCRIPTION
**Pull request by lhecker**
## Summary of the Pull Request

See [my code comment](https://github.com/microsoft/terminal/pull/4150#discussion_r364392640) below for technical details of the issue that caused #4145.

## PR Checklist
* [x] Closes #1360, Closes #4145.
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Detailed Description of the Pull Request / Additional comments

TBH I kinda hope this project could migrate to an internal use of UTF-8 in the future. 😶

## Validation Steps Performed

Followed the "Steps to reproduce" in #4145 and ensured the "Expected behavior" happens.

### CODE REVIEW

File: src/cascadia/TerminalCore/Terminal.cpp [hunk 1]
```diff
@@ -456,21 +456,17 @@ void Terminal::_WriteBuffer(const std::wstring_view& stringView)
         {
             // TODO: MSFT 21006766
             // This is not great but I need it demoable. Fix by making a buffer stream writer.
-            if (wch >= 0xD800 && wch <= 0xDFFF)
-            {
-                const OutputCellIterator it{ stringView.substr(i, 2), _buffer->GetCurrentAttributes() };
-                const auto end = _buffer->Write(it);
-                const auto cellDistance = end.GetCellDistance(it);
-                i += cellDistance - 1;
```
Comment (lhecker) [inline src/cascadia/TerminalCore/Terminal.cpp]: This line causes the loop variable `i` to be decremented if `cellDistance == 0`, which happens to be the case with the HTML from #4145.
Additionally it probably shouldn't have used the `cellDistance` anyways but rather simply increment `i` by 1, whenever 2 code units are extracted from the `stringView`.

**Comment by DHowett-MSFT** [comment]
This has been a historically dangerous area to touch (already!) -- we have had a number of fixes and reversions in this function already.

**Comment by DHowett-MSFT** [comment]
(I think #1360 is the same issue, actually)

**Comment by lhecker** [comment]
You're right - You guys even found the exact same line. 😄

But I believe the conclusion in #1360 was slightly off / incomplete:
Sometimes (?) when you pass 2 code units (a surrogate pair consisting of `[high,low]`), the distance reported after a call to `TextBuffer::Write` will be 1 and not 2. ⚠️
Given a surrogate pair the current code will encounter detect a high surrogate in `wch`, extract 2 code units from `stringView` and give that to `TextBuffer::Write`. Now if `TextBuffer::Write` returns 1, the loop counter `i` will not be incremented.
In the next iteration `wch` will then be the low (⚠️) surrogate of the code point which is fed into `TextBuffer::Write` again, which now returns 0 because the surrogate code unit is incomplete.
Afterwards the loop will now deadlock.

=> @DHowett-MSFT The actual fix AFAICS is to not use `end.GetCellDistance(it)` as a measurement whether `i` needs to be incremented (AFAICS it will only return 1 or 2 anyways).
Instead it should always be incremented by 1 if a surrogate pair is encountered.

**Comment by lhecker** [comment]
Huh... @DHowett-MSFT @miniksa turns out that [the cell distance will always be 1 for surrogate pairs](https://github.com/microsoft/terminal/blob/429af0e6fa80412f1e9185845d856e3497a6cf78/src/buffer/out/OutputCellIterator.cpp#L217-L227)?
When the `OutputCellIterator` is incremented the cell distance (`_distance`) is incremented by 1, whereas the input position (`_pos`) is incremented by 2, thus consuming the entire surrogate pair.
This leads me to believe that `Terminal::_WriteBuffer` probably deadlocks or corrupts most surrogate pairs right now, without this PR.

Here's the before/after with this PR with the string "𐐌𐐜𐐬" (3 surrogate pairs):
image

As you can see the current `Terminal::_WriteBuffer` method currently sadly breaks when faced with surrogates. 😕 This PR mostly fixes it though by not using the cell distance anymore. 🙃

**Comment by zadjii-msft** [comment]
Okay so this has historically been challenging - could we maybe add the tests from #2924 and a test for the #3052 case to this PR? I'd really hate to have another "simple" change in this area secretly bite us again

**Comment by miniksa** [comment]
If we can put some tests in, we can give it another go. Does it also fix #1360, @lhecker?

File: src/cascadia/TerminalCore/Terminal.cpp (hunk 1 above)
Comment (miniksa) [reply src/cascadia/TerminalCore/Terminal.cpp]: Why was `cellDistance == 0`? What was the specific code point that caused the problem?

File: src/cascadia/TerminalCore/Terminal.cpp [hunk 2]
```diff
@@ -456,21 +456,17 @@ void Terminal::_WriteBuffer(const std::wstring_view& stringView)
...
-                const OutputCellIterator it{ stringView.substr(i, 1), _buffer->GetCurrentAttributes() };
-                const auto end = _buffer->Write(it);
-                const auto cellDistance = end.GetCellDistance(it);
-                proposedCursorPosition.X += gsl::narrow<SHORT>(cellDistance);
-            }
+            //
+            // If wch is a surrogate character we need to read 2 code units
+            // from the stringView to form a single code point.
+            const auto isSurrogate = wch >= 0xD800 && wch <= 0xDFFF;
+            const size_t codePointLength = isSurrogate ? 2 : 1;
```
Comment (miniksa) [inline src/cascadia/TerminalCore/Terminal.cpp]: Isn't `codePointLength` always going to be the same value as `GetInputDistance` between the two iterators? Maybe I chose wrong with `GetCellDistance`, but it feels like it was just `GetInputDistance` instead.

File: src/cascadia/TerminalCore/Terminal.cpp (hunk 1 above)
Comment (lhecker) [reply src/cascadia/TerminalCore/Terminal.cpp]: I tried to describe that here: https://github.com/microsoft/terminal/pull/4150#issuecomment-572220629
tl;dr: In an iteration with a surrogate pair `GetCellDistance()` returns 1 (as described below), which causes `i` to be "misaligned" within `stringView`. In the next iteration `TextBuffer::Write` will now encounter a low surrogate pair and either advance the iterator again only by 1 character or fail to compose the character and return 0.

Comment (lhecker) [reply src/cascadia/TerminalCore/Terminal.cpp]: You can see an example of this misalignment quite well in the picture in this comment: https://github.com/microsoft/terminal/pull/4150#issuecomment-572239776
Instead of it printing 3 glyphs it prints 6, due to `stringView` contains 6 code units and `GetCellDistance()` returning 1 at most. (...and never 2!)

File: src/cascadia/TerminalCore/Terminal.cpp (hunk 2 above)
Comment (lhecker) [reply src/cascadia/TerminalCore/Terminal.cpp]: Yes we could use `GetInputDistance()` instead. I checked & it works as well.
Do you want me to use ``GetInputDistance()` instead?

File: src/cascadia/TerminalCore/Terminal.cpp (hunk 1 above)
Comment (lhecker) [reply src/cascadia/TerminalCore/Terminal.cpp]: Oh no! I accidentially deleted one of my comments here. 😱
Do you have emails from GitHub enabled @miniksa? In that case you should find it there. 😣

**Comment by lhecker** [comment]
> Does it also fix #1360, @lhecker?

@miniksa Using [the minimal reproduction program](https://github.com/microsoft/terminal/issues/1360#issuecomment-534310766)? Yup. 🙂
image

File: src/cascadia/TerminalCore/Terminal.cpp (hunk 2 above)
Comment (miniksa) [reply src/cascadia/TerminalCore/Terminal.cpp]: Yeah just because GetCellDistance and GetInputDistance are meaningful sentinels to me of where all in the code someone is doing this type of calculation. (a.k.a. places where this sort of thing goes to hell all the time).

Comment (lhecker) [reply src/cascadia/TerminalCore/Terminal.cpp]: I'll change it. The reason I _personally_ disliked using `GetInputDistance()` until now is because I expect the `TextBuffer::Write` function to consume all given code units in the `OutputCellIterator` and advance `i` accordingly.
This is only the case if you use the same variable you use for `stringView.substr()`, as for incrementing `i`. Again, I'll change it though, because you make a good point.

Comment (miniksa) [reply src/cascadia/TerminalCore/Terminal.cpp]: As a sentinel, if we decide that `OutputCellIterator` is globally stupid for making the caller maintain the walk-count... (which it is... in my opinion...) we can globally change it at some point by finding all affected areas.

I know I don't need to convince you any more, just giving more clarity on my thoughts.

**Comment by lhecker** [comment]
> could we maybe add the tests from #2924

@zadjii-msft I've tried to adapt the referenced test case, but was unable to figure out how make it work in the first place. If I call `Terminal::PrintString()` with a string containing 300 "a" characters it never returns. [snippet](https://gist.github.com/lhecker/96d8d36ebad9536f73d3eddb9873bca9)
I hope I can figure out tomorrow how to do whatever the `runut` CLI does, in Visual Studio so I can debug this better. Otherwise I _might_ have to rely on someone helping me out a bit to adapt #2924's test.

**Comment by zadjii-msft** [comment]
@lhecker ruh-roh
image

I just copy-pasted the test and built it locally, and it passed just fine. So maybe your change _broke_ it 😅 

Usually, when I'm debugging tests, I add a `DebugBreak();` statement and let the post-mortem debugger of WinDbg break into the test process. I don't know if VS has a post-mortem debugger that would work similarly. @miniksa's the other primarily-VS user on the team, so he might have an idea.

**Comment by lhecker** [comment]
> So maybe your change broke it 😅

I tried it on the master branch - even after `bcz dbg`. 🤔 I must be doing something wrong.

> Usually, when I'm debugging tests, I add a DebugBreak(); statement and let the post-mortem debugger of WinDbg break into the test process.

Thank you so much! I'll try fixing it using that tomorrow. 🤗

**Comment by miniksa** [comment]
> Usually, when I'm debugging tests, I add a `DebugBreak();` statement and let the post-mortem debugger of WinDbg break into the test process. I don't know if VS has a post-mortem debugger that would work similarly. @miniksa Michael Niksa FTE's the other primarily-VS user on the team, so he might have an idea.

Go to `Tools > Options` and then make sure that `Native` is checked as the `Just-In-Time Debugging` provider. (Checking the box, if it is not checked, will require that Visual Studio is launched as Administrator.)

image

Then when you run something with `DebugBreak()` in it, you will see this: 
image

The top ones will be new instances of the Visual Studios installed on your system. The bottom ones will be the running instances of Visual Studio. You can see I have one open already. If I choose the bottom one, VS will attach straight up as if I F5'd from the solution at the point from the `DebugBreak()`. Step up once to get out of the break and back into the code.

File: src/cascadia/TerminalCore/Terminal.cpp (hunk 1 above)
Comment (miniksa) [reply src/cascadia/TerminalCore/Terminal.cpp]: I do have e-mails from GitHub enabled, @lhecker, but with a repo as popular as ours, I get something like 3,000 per day. So I can't quite tell which one you deleted, even with threaded mail. :(

Either way, I'm fine with this. I don't need to dig deeper.

**Comment by DHowett-MSFT** [comment]
Here's my trick!

image

image


Setting the command to `te.exe` with the arguments `/inproc $(TargetPath)` lets you run it under debug with <kbd>F5</kbd>. This doesn't work for every test, as some can't run inproc (TE usually spawns a child process to run tests, but `/inproc` makes it run them directly).

You can place and hit breakpoints and everything this way.

**Comment by miniksa** [comment]
> Here's my trick!
> 
> image
> 
> image
> 
> Setting the command to `te.exe` with the arguments `/inproc $(TargetPath)` lets you run it under debug with F5. This doesn't work for every test, as some can't run inproc (TE usually spawns a child process to run tests, but `/inproc` makes it run them directly).
> 
> You can place and hit breakpoints and everything this way.

This is probably another good one for the #4163 treatment.

**Comment by lhecker** [comment]
I added a test. Can you guys re-review? 🙂

---

The long explanation for the actual cause of the freeze:

The strings that get fed into `Terminal::_WriteBuffer` are always chopped into segments of as many characters as the terminal is wide (i.e. 120 chars for a 120 char wide terminal).
But since the string view contains code units and not characters, it'll contain more items than the terminal buffer is wide when surrogate pairs are involved (damn you UTF-16!).
Due to the bug of `GetCellDistance()` being used to increment `i` it'll now loop more than e.g. 120 times (namely twice for every surrogate pair).
This causes the buffer line to be full prematurely and `GetCellDistance()` to return 0.
--> The loop loops infinitely now.
--> Fixed by using `GetInputDistance()` and by adding a safety check ensuring newlines are inserted if a buffer line is already full. 👍

---

@zadjii-msft Thanks for making me write the test. I hardened the code a bit more now!
It also helped me understand the underlying issue and #1360 a bit better (as written above).

<details>
<summary>Regarding DebugBreak</summary>
When I add a `DebugBreak()` statement as the first thing in a `TEST() { ... }` group I get the following error in my terminal:
> A crash with exception code 0x80000003 occurred in module "KERNELBASE.dll" in the process hosting the test code while invoking a test operation

...and no window opens. ¯\\\_(ツ)_/¯
(Because I don't have much free time I just solved it temporarily by adding a `Sleep(10000);` instead and quickly attaching a debugger manually. #lifehacks 😄)
</details>

**Comment by lhecker** [comment]
Hmm... Something strange is still happening. 🤔
If I print a file containing the character "𐐌" ~80 times the following happens in both the classical terminal, as well as the new one:
image

That doesn't happen with emojis. I guess that's another followup ticket?

**Comment by miniksa** [comment]
> Hmm... Something strange is still happening. 🤔
> If I print a file containing the character "𐐌" ~80 times the following happens in both the classical terminal, as well as the new one:
> image
> 
> That doesn't happen with emojis. I guess that's another followup ticket?

Yeah, I'd file it as a follow up. Though I'd be surprised if there isn't a dupe somewhere on the tracker. If it's happening in both, it's almost assuredly a `conhost.exe` bug.

**Comment by lhecker** [comment]
I'm just writing to make sure this PR isn't forgotten. 😄
It only needs one more possible approval to be merged and fixes a "Priority-1" issue. 🙂
